Move to a microservice architecture

ifed3 opened the pull request:
---
name: Upgrade WOAQ infrastructure
about: #122
title: Set up microservices
labels: infrastructure, api, web, dataviz
reviewers: @r-b-g-b, @theecrit, @nbilenko, @TangoYankee 

---

## Checklist
- [X] Add description
- [X] Reference open issue pull request addresses
- [X] Pass linting check
  - complete on the local machine with `make local.shell` `make quality`
- [X] Pass functional tests
  - complete on the local machine with `make local.shell` `make test`
- [X] Request code review
  - Please allow **36 hours** from opening a pull request before merging a pull request- even if it has already recieved an approving review.
- [x] Address comments on code and resolve requested changes
- [x] Merge own code

## Description
Issue: #122 

Thanks for your patience. This PR is finally ready.

* There are now five services in the WOAQ architecture: django rest api, react client, jupyter notebook for dataviz, postgis database, reverse proxy with automatically generated ssl certificates
* Services are no longer accessible by ports, instead they are defined by domains
    * Web interface is available on [lvh.me](http://lvh.me)
    * API is available on [api.lvh.me](http://api.lvh.me)
    * Data visualization notebook is available on [dataviz.lvh.me](http://dataviz.lvh.me)
* README has been cleaned up to make it easier to get up and running. Project overview and historical info have been moved to the About.md section
* Python dependency manager has been changed from pipenv to [pip-tools](https://github.com/jazzband/pip-tools). It's simpler for working with Docker environments. Read more [here](https://pythonspeed.com/articles/pipenv-docker/)
* Make commands that relate to docker have been removed due to refactored docker-compose files
and multiple services instead of a single one
    * Run `docker-compose up -d --build` instead of `make local.build` + `make local.up` to build and run all services
    * Run `docker-compose up -d --build service_name` to run a specific service
    * Run `docker-compose down` instead of `make local.down`
    * Run `docker-compose run service_name /bin/bash` instead of `make local.shell` to enter the shell of a specific service
* One thing to pay attention to is the change in directory structure. All services have their own directory. Make sure to `cd` into the correct directory if you want to run commands for a specific service

p.s there are still some outstanding issues relating to devops/infrastructure that will need to be completed in the near future: writing tests for the client service, creating a Heroku container configuration file, setting up the Heroku deploy process, pushing container images to Docker hub to enable faster CI builds

ifed3 committed:
Remove terraform linked files

ifed3 committed:
Move woeip directory to api folder

ifed3 committed:
Simplify travis CI configuration

ifed3 committed:
Use pip-tools as dependency manger instead of pipenv

ifed3 committed:
Update docker ignore files

ifed3 committed:
Edit proxy config to use api service

ifed3 committed:
Remove static file and log directory from start commands

ifed3 committed:
Set top level env file to handle python environment

ifed3 committed:
Move python environment to env file within api directory

ifed3 committed:
Use multi stage docker build with alpine base for api image

ifed3 committed:
Clean up build process

ifed3 committed:
Install razzle with typescript for react client usage

ifed3 committed:
Move volume defintion

ifed3 committed:
Set up web client service

ifed3 committed:
Fix typescript signature match error

ifed3 committed:
Create Makefile for docker commands within root

ifed3 committed:
Add command to build a specific service

ifed3 committed:
Fix web dockerfile creation

ifed3 committed:
Create separate docker configs for web in dev and prod

ifed3 committed:
Use nginx reverse proxy locally to allow for domain functionality

ifed3 committed:
Use full web volume instead of copying files

ifed3 committed:
Enable HMR when using a docker container

ifed3 committed:
Mirror django update with pip-tools instead of pipenv

ifed3 committed:
Merge branch 'master' into ia/set-up-microservices

ifed3 committed:
Remove nginx config since proxy docker container is auto configured

ifed3 committed:
Set explicit build env var in docker compose files

ifed3 committed:
Allow cors for api domain

ifed3 committed:
Move new schema migration files to correct directory

ifed3 committed:
Remove base makefile in favor of docker-compose commands

ifed3 committed:
Simplify api start script

ifed3 committed:
Update django to v2.1.11

ifed3 committed:
Simplify readme and move project details to about section

ifed3 committed:
Remove static declarations from api environment

ifed3 committed:
Update contributing document

ifed3 committed:
Set up data visualizations service

ifed3 committed:
Add secure environment variables to travis

ifed3 committed:
Change docker version to allow travis build

ifed3 committed:
Update travis config and create local compose file

ifed3 committed:
Update docker compose version

ifed3 committed:
Add python env var to travis

ifed3 committed:
Fix travis env variables

ifed3 committed:
Change application name to api for travis build

ifed3 committed:
Fix db hostname to enable migrations to run

ifed3 committed:
Refactor docker-compose files for simpler usage

ifed3 committed:
Fix missing docker command

ifed3 committed:
Add bash command to run travis scripts

ifed3 committed:
Add command flag to travis bash scripts

ifed3 committed:
Set api CI entrypoint to be bash to allow script commands to be run

ifed3 committed:
Wait for database to be ready before running validation scripts

ifed3 committed:
Remove entrypoint from local docker-compose config

ifed3 committed:
Remove pylint and pytest from being ignored in container

ifed3 committed:
Refactor dustrak and gps file loading and remove localization since pandas is now tz-aware

ifed3 committed:
Re-add gps log file to repo

ifed3 committed:
Attempt to get code coverage working on travis

ifed3 committed:
Run only xml coverage instead of report coverage when testing

ifed3 committed:
Run code coverage as part of the CI build process

ifed3 committed:
Move codecov to run in docker container

ifed3 committed:
Move codecov install to docker container

ifed3 committed:
Create artificats folder to store coverage xml report

nbilenko commented:
Lots of awesome work here @ifed3. The infrastructure work looks great. The cleaner directory structure and separate Dockerfiles make it much easier to understand the codebase and to isolate changes related to different microservices.

However, there are changes that are out of scope of the issue, and this makes this PR quite large and more difficult to review. I would suggest retaining the work related to setting up microservices in this PR and separating out the other changes into different issues/PRs.

Some examples of changes I would suggest relocating:
- Python dependency management. I would be totally open to a change, but it's out of scope of this issue.
- Removal of `make` commands. Retaining `make` commands shouldn't impact anything about `docker-compose` commands since they are just aliases. It also seems unrelated to the issue.
- Major README cleanup. The cleanup looks great! But general cleanup for easier onboarding should be a separate issue. The changes in this PR should only be related to the microservice related changes.
- Adding a jupyter notebook. I would be concerned about running a jupyter notebook in production because of how computationally expensive it would be. Also, notebooks are great for interactive data analyses, but very bad for displaying pre-made visualizations. In any case, this change is out of scope of this issue.

Since I'm asking for a major PR change, I'll hold off on specific review of the infrastructure changes, but I'm looking forward to it - seems like a great way to move forward.

ifed3 committed:
Use the dataviz service only for the local dev env

ifed3 committed:
Condense validate script to include migration, quality checks and tests

theecrit commented:
I agree with @nbilenko here; we really want to make sure our issues contain discrete implementations so work can be tracked more easily and discussions can remain manageable. The issue breakdown provided feels appropriate to me, too.

Thanks,
J.

ifed3 committed:
Copy coverage file from docker container to travis host to run codecov

ifed3 commented:
Good catch @nbilenko with the dataviz service running in production. I've updated that.

I think it would be misguided to say that changes are out of scope of this refactor. They are interwoven in a manner that would require more work to separate them out in a new PR.

My aim was to build an architecture that can be run locally and on production without major configuration, so with microservices, three high level changes happened:
1. The only supported way to run the infrastructure is via docker, i.e you can't run it directly on your host machine
2. The only environment to run locally via docker is the dev/test environment. There will never be a need to run the prod or CI env locally
3. Subdirectories now exist thus each directory should only contain files relevant to its service

## Switch from pipenv to pip-tools
pipenv by default provides a virtual env, serves as both a dependency manager and installer. This is ideal if you want to run a python app on your host machine but is problematic for docker usage.

pipenv uses different install commands based on the env: `pipenv install --dev` vs. `pipenv install`. To solve this would require creating a different Dockerfile to run the api service in prod simply due to one command change. Or to create a conditional bash script to check the env and run the appropriate command.

Once there is no need to run the infra on your host machine, it actually becomes both unnecessary and more cumbersome to maintain support for pipenv.

## Makefile docker-compose aliases
Once the django app is moved to the api directory, it takes the `Makefile` with it since most of the commands are python/django related.

To keep the aliased docker commands would require creating a new Makefile within the root directory. The aliases were ideal when you could choose an env to run (local vs. production) but became redundant when you're limited to only run the local env

So you go from:
```bash
make %.up: ## Start the (local|production) Docker containers
    docker-compose -f docker-compose.yml -f docker-compose.$*.yml up -d
```
to: 
```bash
make local.up ## Start the local Docker containers
    docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d
```

However by default, docker-compose searches for two config files (`docker-compose.yml, docker-compose.override.yml`) when running; once I renamed `docker-compose.local.yml` to `docker-compose.override.yml`, these commands all became equivalent!
```bash
make local.up
docker-compose -f docker-compose.yml -f docker-compose.override.yml up -d
docker-compose up -d
```

Thus all the aliases essentially became plain and easy to use docker-compose commands and getting familiar with docker-compose syntax will be far more important for one's coding chops than using arbitrary aliases especially for such basic commands. 

So the aliases became counterproductive both from a code functionality and developer usage standpoint with a microservices architecture

## README/contributing.md
These changes relate to the onboarding and use of the micoservices architecture. If I had left them as they were, half of it would be irrelevant and non-functional since it assumes just a django app rather than multiple apps.

## Holoviews
This is the change I agree is extraneous and probably out of scope. 

To your point, @nbilenko, Holoviews was built specifically to solve the problem of displaying pre-made visualizations. Read more [here](http://holoviews.org/user_guide/Plots_and_Renderers.html). It can output static displays in many formats such as `png`, `jpeg`, `svg`, `html` and dynamic displays as `html` + `js`. because it uses [Bokeh](https://bokeh.pydata.org/en/latest/) as its backend provider which is described as follows:

```
Bokeh is an interactive visualization library that targets modern web browsers for presentation. Its goal is to provide elegant, concise construction of versatile graphics, and to extend this capability with high-performance interactivity over very large or streaming datasets
```

----
Hope that explains some of the changes. I think this conversation would be best be expounded upon and resolved in person. See you all tonight!

ifed3 committed:
Run coverage copy within same VM as coverage creation

ifed3 committed:
Fix missing container name issue

nbilenko commented:
> I think it would be misguided to say that changes are out of scope of this refactor. They are interwoven in a manner that would require more work to separate them out in a new PR.

The interweaving of the changes is exactly what my feedback is about - the interweaving complicates both the review and keeping track the history of these changes. The different changes I mentioned are hard to delineate and to keep track of while reading the code. Also, discussion for all changes has to be maintained in a single comment thread. E.g. a chat about pros and cons of dataviz software sounds great, but an infrastructure PR comment thread is not where decisions should be made about what software the team will be using for data visualization.

I realize it would take some work for you to separate out the changes, but to read and understand these interwoven changes would instead place a substantial amount of work on the rest of the team. It would also likely to not lead to effective review - best practices tend to recommend limiting code changes to a few hundred lines because reviewers can't reliably detect problems with longer change lists.

Let's discuss more tonight.

ifed3 committed:
Fix coverage file path

ifed3 committed:
Remove data visualization service/container

ifed3 committed:
Move dev environment variables from env file to compose section

ifed3 committed:
Bind guincorn back to port 8000

ifed3 commented:
hey all,

This is the new [infra document](https://user-images.githubusercontent.com/9001622/63615519-6b755d80-c59a-11e9-8fe8-2cab84eae2e7.jpg)

Here's a more thorough breakdown of the changes that happened within this PR. If you have any questions or any confusion, please let me know.

Docker
---------
- Remove all files related to AWS, Terraform, Elastic Beanstalk
- Move django files to api directory
- Create react client service within web directory
- Replace nginx service and configs with fully managed nginx-proxy service
- Use nginx-proxy to map ports to domains
- Set up CORS configuration for nginx-proxy so api domain can be accessed by any other domain
- Use nginx-proxy-letsencrypt-companion to automagically set up SSL certificates for domains
- Rename docker-compose.local.yml to docker-compose.override.yml
- Rename docker-compose.travis.yml to docker-compose.ci.yml to align with environment
- Create docker-compose.prod.yml to set up production environment
- Remove docker-compose commands from Makefile due to django directory change and renaming of local compose override file
- Update dockerignore files to minimize unncessary files being copied to containers
- Drop pipenv to avoid api Dockerfile duplication for multiple environments
- Remove pipenv environment variables from api Dockerfile
- Remove logging configuration in api Dockerfile and entrypoint script in favor of PYTHONUNBUFFERED env variable and nginx-proxy taking care of gunicorn logs 

CI (Travis)
----------
- Replace wait-for-it script with dockerize library to avoid mantainance issues
- Await database connection with dockerize wait command to prevent CI build/test failure
- Setup secure environment variables to be used on Travis CI
- Simplify CI build and test scripts
- Update code coverage script

Wiki
-----------
- Clean README to set focus on awareness of available services and getting started quickly
- Move project details and history to About section
- Update contributing section to remove instructions on running on host machine (now only supported via docker-compose)
- Update contributing section to remove details about logging in to django admin (not needed for api usage)
- Update contributing section to provide a clear step-list for getting from issue to pull request
- Update contributing section to provide a one-liner for testing the api service without using aliases

Python
---------
- Replace Pipfile/Pipfile.lock with requirements.in/requirements.txt 
- Set python version with runtime.txt since Pipfile no longer exists
- Fix failing test from get_target_data method due to new pandas lib version being timezone aware
- Fix failing test from test_joiner and load_dustrak methods due to complex newline split script

nbilenko commented on docker-compose.override.yml:
```diff
@@ -0,0 +1,59 @@
+      VIRTUAL_HOST: api.lvh.me
+      VIRTUAL_PORT: 8000
+      SECRET_KEY: secret_key
+      DEBUG: true
```
This has to be in quotes to get passed as an env variable to Python b/c otherwise yaml treats it as a boolean:
`true` => `'True'`
C.f. https://github.com/docker/compose/issues/1788

nbilenko commented on docker-compose.override.yml:
```diff
@@ -0,0 +1,59 @@
+      SECRET_KEY: secret_key
+      DEBUG: true
+      DATABASE_URL: postgis://postgres:postgres@db:5432/woaq
+      ENABLE_LOGGING_TO_FILE: false
```
This has to be in quotes, same as above:
`false` => `'False'`

nbilenko commented on README.md:
```diff
@@ -1,51 +1,50 @@
-We use [Trello](https://trello.com/b/EBnxZHmx/west-oakland-air-quality) to document and discuss data decisions:
-- Review and add a comment to any yellow “Data Q” card currently sitting in the [Qs About Data Analysis](https://trello.com/b/EBnxZHmx/west-oakland-air-quality) column. Or;
-- Add a new yellow “Data Q” card to the [Qs About Data Analysis](https://trello.com/b/EBnxZHmx/west-oakland-air-quality) column for any new questions you think of that might be important to answer. 
+Browse the client at [lvh.me](http://lvh.me), ping the api at [api.lvh.me](http://api.lvh.me).
```
Suggestion: add a line about how to bring the service back down.

nbilenko commented on README.md:
````diff
@@ -1,51 +1,50 @@
+```bash
+  git clone https://github.com/openoakland/woeip.git
+  cd woeip
+  docker-compose up -d --build
````
Since the db name has changed, I had to create a new database `woaq`, so I would suggest adding that in the README as an instruction for initial setup. This was the way I did it:
```
docker exec -it woaq-db /bin/bash
psql -U postgres
CREATE DATABASE woaq;
```

r-b-g-b replied:
Talking to @ifed3 now about this. When I set this up fresh, I didn't have to create the database manually, but worth looking into what was different in your case. One solution would be to have an entrypoint script that does this.

r-b-g-b replied:
Opened #126 to address.

nbilenko reviewed (changes requested):
Thank you for the detailed breakdown of the changes and documentation, Ife! I added inline comments about changes I had to make to get things running. Overall, looks great to me.

r-b-g-b commented on docker-compose.yml:
```diff
@@ -1,32 +1,29 @@
-  logs:
-  media:
-  static:
+      - /var/run/docker.sock:/tmp/docker.sock:ro
```
Just curious: Is there a simple explanation for why we link the docker socket like this?

r-b-g-b commented on web/.dockerignore:
```diff
@@ -0,0 +1,11 @@
+.dockerignore
+.env
+.env.example
```
Do we want to ignore `.env.example` too?

r-b-g-b reviewed (approved):
Very cool @ifed3 ! Thanks for this. I learned a lot from reviewing this -- [extending docker-compose files](https://docs.docker.com/compose/extends/), [dockerize](https://github.com/jwilder/dockerize), [pip-tools](https://github.com/jazzband/pip-tools).

I did my best to understand the web service, SSL, letsencrypt, etc., but can't claim to be very knowledgeable about that, so I'll trust that you made it work. At the very least, I was able to get it up and running using the nice docs you provided.

I second @nbilenko 's point about preferring smaller PRs, so let's keep that in mind in the future. Approved!

r-b-g-b committed:
Put true/false in quotes so yaml doesn't treat as bool
